src/main/java/org/example/controllers/WeatherController.java

```diff
@@ -0,0 +1,39 @@
+package org.example.controllers;
+
+import org.example.models.dto.WeatherDTO;
+import org.example.services.WeatherService;
+import org.springframework.http.ResponseEntity;
+import org.springframework.validation.annotation.Validated;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@Validated
+public class WeatherController {
+
+    private final WeatherService weatherService;
+    public WeatherController(WeatherService weatherService) {
+        this.weatherService = weatherService;
+    }
+
+    @PostMapping("/api/weather")
+    public ResponseEntity<WeatherDTO> storeWeather(@RequestBody @Validated WeatherDTO weatherDTO) {
```

Comment:
If we want to make validations on DTOs, we can use the `@Validated` annotation on the controller class as you did with `WeatherController`, but in the parameter's controller method, we can/should use `@Valid` annotation. Having that in place, we can further use `@NotNull`, `@NotEmpty`, etc annotations on DTOs fields as desired.

You can check [this](https://www.baeldung.com/spring-boot-bean-validation) example. Observe the `@Valid` annotation from the controller class, and annotations from the entity, but in our case, we want to make that validation on the DTO fields.
Please ignore the ExceptionHandler or the testing part from the example.

Reply:
Hey Alex, I've added the spring boot starter validation in the pom.xml and validated the fields of the WeatherDTO, yes indeed it seems I got it wrong when it came to the validation on the parameter's methods in the controller class. I have changed it from @Validated to @Valid. The fields from the DTO have been validate with the @NotBlank on the city String DataType and @Max and @Min on the temperature and humidity.